source/opt/ssa_rewrite_pass.cpp

```diff
@@ -0,0 +1,481 @@
+
+#if SSA_REWRITE_DEBUGGING_LEVEL > 0
+    std::cout << "\tFound load: "
+              << inst->PrettyPrint(SPV_BINARY_TO_TEXT_OPTION_FRIENDLY_NAMES)
+              << " (replacement for %" << load_id << " is %" << val_id << ")\n";
+#endif
+  }
+}
+
+void SSARewriter::GenerateSSAReplacements(ir::BasicBlock* bb) {
+#if SSA_REWRITE_DEBUGGING_LEVEL > 0
+  std::cout << "Generating SSA replacements for block: " << bb->id() << "\n";
+  std::cout << bb->PrettyPrint(SPV_BINARY_TO_TEXT_OPTION_FRIENDLY_NAMES)
+            << "\n";
+#endif
+
+  // Seal |bb|. This means that all the stores in it have been scanned and it's
+  // ready to feed them into its successors. Note that we seal the block before
+  // we scan its instructions, this way loads reading from stores within |bb| do
+  // not create unnecessary trivial Phi candidates.
+  SealBlock(bb);
+
+  for (auto& inst : *bb) {
+    auto opcode = inst.opcode();
+    if (opcode == SpvOpStore || opcode == SpvOpVariable) {
+      ProcessStore(&inst, bb);
+    } else if (inst.opcode() == SpvOpLoad) {
+      ProcessLoad(&inst, bb);
+    }
+  }
+
+#if SSA_REWRITE_DEBUGGING_LEVEL > 0
+  std::cout << "\n\nPhi candidates:\n";
+  for (const auto& phi_it : phi_candidates_) {
+    for (const auto& it : phi_it.second) {
+      std::cout << "\tBB %" << phi_it.first->id() << ": "
+                << it.second.PrettyPrint(pass_->cfg()) << "\n";
+    }
+  }
+
+  std::cout << "\nLoad replacement table\n";
+  for (const auto& it : load_replacement_) {
+    std::cout << "\t%" << it.first << " -> %" << it.second << "\n";
+  }
+
+  std::cout << "\n\n";
+#endif
+}
+
+bool SSARewriter::ApplyReplacements() {
+  bool modified = false;
+
+#if SSA_REWRITE_DEBUGGING_LEVEL > 0
+  std::cout << "\n\nApplying replacement decisions to IR\n\n";
+  std::cout << "Converting non-trivial Phi candidates into Phi instructions\n";
+#endif
+
+  // Add Phi instructions from all Phi candidates.
+  for (auto& it : phi_candidates_) {
+    ir::BasicBlock* bb = it.first;
+
+#if SSA_REWRITE_DEBUGGING_LEVEL > 1
+    std::cout << "\nNon-trivial Phi candidates at block %" << bb->id() << "\n";
+#endif
+    const auto& phis_at_block = it.second;
+    auto insert_it = bb->begin();
+    for (auto& phi_it : phis_at_block) {
+      const PhiCandidate& phi_candidate = phi_it.second;
+
+#if SSA_REWRITE_DEBUGGING_LEVEL > 1
+      std::cout << "Phi candidate: " << phi_candidate.PrettyPrint(pass_->cfg())
+                << "\n";
+#endif
+      if (phi_candidate.is_trivial) {
+        continue;
+      }
+
+      uint32_t type_id = pass_->GetPointeeTypeId(
+          pass_->get_def_use_mgr()->GetDef(phi_candidate.var_id));
+      std::vector<ir::Operand> phi_operands;
+      uint32_t ix = 0;
+      for (uint32_t pred_label : pass_->cfg()->preds(phi_candidate.bb->id())) {
+        uint32_t op_val_id = phi_candidate.phi_args[ix++];
+        phi_operands.push_back(
+            {spv_operand_type_t::SPV_OPERAND_TYPE_ID, {op_val_id}});
```

Comment:
We already identified that this is a potential problem because if `op_val_id` is the result of a trivial phi candidate, the instruction defining this id will not be generated.

The solution I propose is a larger change but it can work:

1) Change `phi_candidates_` to a map from the result id the of candidate to a `PhiCandidate`.  It does not seem like the map from the basic block is actually used anyway, so it should not have a big impact on existing code.

2) At this point add code that as long as `op_val_id` is the potential result of a trivial phi will set it to the single phi_arg.   Then generate the instruction using that value instead.

The same change will be needed when replacing loads below.

Reply:
This has changed quite a bit now. PTAL.